### ISSUE TITLE
RFC: Switch statement

### ISSUE THREAD

**Issue by elliottslaughter** [terralang/terra#417]
This is new for us: I'm going to try to write out an RFC for a proposed syntax change in #416, which adds a `switch` statement to the language. The goal here is to get the syntax changes all in one place so that there is a definitive definition of what we're trying to do (and so you don't have to go chasing comment threads to track the latest state). I will edit this RFC as we make changes.

Bear with me since this is new and we may evolve the process as we go along.

# Motivation

Switch is a useful construct to have in a low-level language like Terra because it exposes a low-level machine feature (jump tables) in a zero-cost way that otherwise you would not be able to hand-code without language support. The initial support suggested here is fairly minimal (i.e. we're not going to full-on Rust-style `match`) which makes it a good fit for Terra's low-level optimization goals. Higher level constructs can be metaprogrammed, or possibly considered as future additions to the language.

# Syntax Proposal

This proposal adds the following syntax:

```
switch value_expr do
  case case_const then
    case_body
  case case_const then
    case_body
  ...
else
  else_body
end
```

The `else` is optional and provides a default block that runs if no case matches. `case_const` are constant values (currently just integers; see more below). `*_body` are blocks of statements like you'd have in any control flow construct. `value_expr` is an arbitrary expression of the same type as the case constants.

# Case Constants

The `case_const` above must be compile-time constant values. These values may be metaprogrammed either explicitly or implicitly. Currently only integers are allowed. The intention is that this can be extended in the a forwards-compatible way in the future.

The following are all legal:

```
local x = 3
terra test_valid_case()
  switch ... do
    case 1 then ...
    case [1+1] then ...
    case [x] then ... -- note the definition of `x` above
    case x then ... -- also ok without the `[]`
    end
  end
end
```

The following are **not legal**:

```
terra test_invalid_case()
  var y = 4
  switch ... do
    case y then ... -- error: y is a Terra variable
    end
  end
end
```

# No Default Fallthrough

This proposal does **NOT** provide default fallthrough between case statements, unlike C. Therefore, there is no need to use `break` or `return` to avoid fallthrough.

If we provide fallthrough in the future it will be via explicit keyword (to be determined). Therefore code written for today's semantics will work on future versions even if fallthrough is later added.

# Variable Scoping

Each case block is its own scope, just like normal Lua control flow constructs. This means in the future if we support fallthrough, if the user wants variables to be visible across multiple blocks they will need to declare then in an outer scope (just like chained `if`/`else`).

# Case Block Nesting Requirements

Case blocks must be nested directly under the `switch` statement. That is, the following is **not legal**:

```
switch ... do
  while ... do
    case ... then
      ...
```

This (intentionally) prevents the use of Duff's Device. The plan is to enable Duff's Device through a separate feature (either jump tables or computed goto), though the current proposal does not prevent this from being added directly to switch statements in the future if desired.

# Metaprogramming

The switch statement can be metaprogrammed by putting standalone case blocks inside quotes. Note that such quotes can only be expanded directly into switch statements, as per the requirements above.

The following is legal:

```
local x = quote
  case 1 then ...
  end
end
terra test_valid_metaprogramming()
  switch ... do
    [x]
  end
end
```

# Alternatives

Another option is to omit `do` after the switch expression. This streamlines the syntax somewhat, but it makes switch a little inconsistent with other control flow constructs, and also introduces hazards around the use of the escape operator `[]`.

# Open or Previously Discussed Questions

  * What values can we match on? The current implementation only does integers, but it may be worth defining in this RFC a larger set of values that we expect to eventually be supported.
  * Do `case` blocks define their own scopes? This is relevant because it influences variable lifetimes in the event of fallthrough. I've suggested one option above but this is open to discussion.
  * Does `case` require its own `end` (seperately from the top-level `switch`)? Currently I'm leaning toward yes because otherwise it makes the metaprogramming ugly (i.e. when you need to put a `case` directly inside a quote).

**Comment by aiverson** [comment]
If we have fallthrough, the extra do at the beginning would be essential for rendering things like duffs device that includes a loop header between the switch statement and the first case.

**Comment by aiverson** [comment]
With regards to allowing things other than integers, having it turn into integers is important to being able to render it as a branch table in assembly for performance reasons. It could potentially be extended with metamethods like __for to allow custom matching behavior, or even to allow pattern matching with algebraic data types, but I don't think we have a good way to do that at all yet. The important thing is that it efficiently compiles into a branch table natively, and the current implementation does everything needed for what I and @blackhole12 are working on.

**Comment by blackhole12** [comment]
Cases currently define their own scopes in the given implementation. This is intended, because cases in C++ do _not_ define their own scopes and this is legendary for how annoying it is. Almost all modern languages deliberately have cases define their own scopes for the same reason they do not have default fallthrough behavior.

**Comment by elliottslaughter** [comment]
> things like duffs device that includes a loop header between the switch statement and the first case

Could you maybe explain with a short example? I'm not sure I understand.

**Comment by aiverson** [comment]
[Duff's device](https://en.wikipedia.org/wiki/Duff's_device) is a method for optimizing a loop unrolling in C by using a switch statement to jump into the middle of a loop that then runs normally via case fallthrough.

Basically, it works by entangling the syntax of a loop and a switch in an awkward way so that the case lables point to partway through the loop. It is a loop unrolling optimization and usually speeds up code by reducing the number of branches in a loop, and by using the same loop body for both the long and short passes it reduces code size and number of branches over an implementation that has separate large and small loops or a large loop and a separate switch. Hardware magic with branch predictors and compiler details may or may not make this setup faster, but it is sometimes useful for things like unrolling a loop the compiler doesn't know how to unroll automatically.

Here is a basic sketch of how that might appear in Terra. (untested code, assuming the use of the `fallthrough` keyword rather than the other option `continue`, parser doesn't currently understand this)
```terra
local function duff8(count, body)
  return quote
    var count_ = [count]
    var idx = 0
    switch (n and 7) do
      while count > 0 do
        escape
          for i = 0, 8 do
            local offset = (8 - i) % 8
            emit quote case [offset] then [body(`[i]+idx)] fallthrough end
          end
        end
      end
    end
  end
end

local terra memcpy(dst: &opaque, src: &opaque, len: intptr)
  [duff8(`len, function(idx)
    return quote
      [&uint8](dst)[idx] = [&uint8](src)[idx]
    end
  )]
end
```

**Comment by aiverson** [comment]
Also, making the case support fallthrough in a way that enables the example up there and create lexical scopes would be difficult. @blackhole12 I have honestly never heard of the lack of case scoping in C++ being legendarily annoying, could you elaborate on that?

**Comment by blackhole12** [comment]
@aiverson [Behold](https://www.google.com/search?q=variable+initialization+skipped+by+case+label&oq=variable+initialization+skipped&aqs=chrome.1.69i57j33.5303j1j7&sourceid=chrome&ie=UTF-8). The fact that you cannot declare a variable inside a case statement is by far one of the most unintuitive things about the C++ switch statement, even more unintuitive than the default fallthrough behavior. No other language I know of has this behavior, because all other languages create scopes for each case label.

**Comment by aiverson** [comment]
Huh. I guess there are a lot of people confused by that, and I can see myself making that mistake, but not being very confused by why it happened. I agree that making it more intuitive and less confusing is a good idea, and that is a good way to do it.

**Comment by ErikMcClure** [comment]
I would like to push this syntax change through. I believe aiverson's example of duff's device demonstrates how metaprogramming code could make a switch statement that does not have `do` ambiguous. For the sake of being conservative about what is or isn't ambiguous, I would recommend keeping `do` to avoid shooting ourselves in the foot and allow us to move forward on the RFC. If we discover at a later time that `do` is unnecessary, we can make it optional. 

If we can agree on the final form of the _initial_ syntax, we can then merge #416 once it passes the tests (or disables the leak test, which does not seem to be functioning correctly).

**Comment by elliottslaughter** [comment]
I've been trying to think through whether there might be an alternative way to enable duff's device without entangling different control flow constructs, which I personally consider sort of an anti-feature of C. For example, would it be sufficient to use `goto` and labels, if we also added a first-class way to do jump tables? E.g. `x = terralib.jumptable(list_of_keys, list_of_labels)` or something like that, and then `goto [x]` or something similar becomes the equivalent of a `switch` statement's jump.

But at any rate, I agree `do` is a more conservative initial syntax. Though I'm not sure I'd want to make the `do` optional later since optional things clutter the syntax too, so it's not entirely without cost to go with it.

I'd still like to get @zdevito to look at this if possible since it is a syntax change.

**Comment by OvermindDL1** [comment]
Case's in C are just `goto`'s under the hood (that's why they are scope-free), so yes label's and goto's can replicate a DUFF's device trivially, like this in psuedo-code (pretend this supports var jump labels like GCC does for C or some other languages):
```c
TOP: goto [J7, J6, J5, J4, J3, J2, J1, J0][size(input)%8];
J7: parse(input[7]);
J6: parse(input[6]);
J5: parse(input[5]);
J4: parse(input[4]);
J3: parse(input[3]);
J2: parse(input[2]);
J1: parse(input[1]);
J0: parse(input[0]);
if(size(input)>7) goto TOP;
```
This is actually pretty close to how a case statement is compiled (an array of jump labels that are indexed in to).

**Comment by aiverson** [comment]
I'm in favor of having a jumptable construct. That would allow making state machine compilers much more easily and more efficiently. If we have the combination of a switch statement and/or match expression/statement that doesn't allow fallthrough, that does have separate scopes for each case, that doesn't allow entangling different control flow constructs (but does still allow using an escape to generate the cases), and that does have some form of metamethod override like the range based for loops, that fits every usecase I want out of this thing very neatly and sidesteps some of the implementation problems regarding how scopes are handled and where they begin and end.

**Comment by zdevito** [comment]
It would be helpful to have some elaboration on the semantics. In particular, what can appear in
a case_const expression? 'just integer' does that mean it only the literal integer? Can it be
a variable that resolves to integer in Lua? Can it be escaped? Is it implicitly escaped like type expressions?
IIRC this is a new concept: Terra otherwise does not have a concept of a constant expression other than
an escaped Lua expression, so it is probably not a great idea to add a new concept for once new 
control flow construct. Implicitly escaping the expression after the case is appealing, but it may 
be pretty misleading.

Terra is primarily a language designed to be meta-programmed, so maybe this can also clarify 
how someone will meta-program a switch statement. For instance, I have a list of generated keys,
and generated code to implement an interpreter loop, how do I generate the switch construct?
It would be better if we didn't have to introduce _new_ meta-programming construct to meta-program
a switch statement, and having it be impossible to meta-program is probably a no go. Considering
how rare a switch statement actual is, it would be better to compromise on the syntax than
to compromise on the meta-programmability.

Fallthrough vs No: I think the C family is unique in fallthroughs and as others have commented
it doesn't work well with when variables go into/out of scope. So a switch statement probably
doesn't need fallthrough. However, computed goto _is_ something that is useful, especially when
trying to generate code. Consider something like:

```
jumptable(x, [3, 4], [::for_three::, ::for_four::], ::for_default::)

::for_three::
...
::for_four::
...
::for_default::

```

Since the things in jump tables args are just lists of things, then it is more clear how they can 
be meta-programmed. In this world, do we need both a switch statement and a jump statement?
Or is a comptued jump sufficient?

**Comment by aiverson** [comment]
From the perspective of the compiler, a splice of a value and a literal that isn't spliced are the same. They generate exactly the same AST, IIRC. Terra does actually have a concept of a constant expression, featuring a core library function to create a constant, and a method on quotes to reduce it to a constant value or check that it is. In my goal implementation, which isn't necessarily what is actually implemented yet, the expression of the switch can be an expression with metaprogramming allowed. The body of the switch is just a list of cases, and the entire thing can be replaced with a splice containing code to generate the cases and a quote containing a single case is allowed. The individual cases values can be any expression that can be coerced to a constant by the built in constant call, including expressions and metaprogramming.

Technically, everything that this switch statement would do could also be done by directly manipulating jump tables and labels, but labels and jump tables are more verbose and easier to mess up. They're good for taking precise control of low level compilation of a language written in Terra or a very large macro that features a few things that switch statements wouldn't be good for, but not good for just ordinary code or macros. The switch syntax makes it more concise and legible and is therefore easier to use most of the time.

**Comment by OvermindDL1** [comment]
I'm also for not having a switch statement, but am for having the primitives necessary to build one.  We don't even need a jump table, we just need a `goto` that can jump to an arbitrary address, and a way to get an address from a label, these are built-in GCC intrinsics for example that allow for building even more optimized jump code than even a jump table can do.  For example see the cpython interpreter compiled with or without the GCC intrinsics (a #DEFINE you can set) and it's quite a substantial speed difference, the intrinsics uses a jump-to-address via labels-converted-to-addresses-at-compiletime and the no-intrinsic version uses a giant `switch`.

**Comment by elliottslaughter** [comment]
I think there may still be a case for a switch statement (for ease of use in the simpler cases), but I agree that it may be better to start with jumptables or computed goto, in part because the integration with metaprogramming is more obvious, and also because there are fewer degrees of freedom in the design so it's more obvious when we've got it right. We can always add switch statements later as a sugar for these other forms.

@OvermindDL1 can you talk more about what performance benefit you get from working directly with label addresses? I'd ideally like to see a performance demonstration before we make any commitment to support this.

**Comment by OvermindDL1** [comment]
With a switch case the input data needs to be calculated and hashed the right way to fit the jump table, which also requires loading the jump table code into memory (as instructions), where with a calculated jump, python for example holds the address that needs to be jumped to directly next (it precalculates the instruction opcode's to calculated jumps on load so no further calculations are necessary).

**Comment by aiverson** [comment]
Point of fact, we can't actually make a switch statement in the form of a macro. The syntax of macro calls prevents having a statement block in the argument which is why I asked #363 to help resolve that. More to the point, Terra isn't lisp and should avoid the problems with lisp. Deeply nesting parenthetical calls that act like macros or functions are much harder to read and can't give as good of syntax errors. Making switch statements be a macro with an extraordinarily awkward syntax would be terrible for the usability of the language. Switch/match is a sufficiently important part of the language that it should have dedicated keywords, syntax, and good error messages rather than an obscure, hard to read and interpret macro call.

Computed goto does seem like a useful construct to have for optimization and implementation, however, type safety and implementation are a major problem, because it breaks assumptions about the structure of the program that LLVM bitcode relies on and debugging operations use as well. The "address of label" construct must return a value of a type that cannot have arithmetic performed on it, but can only be used as constants in an initializer. However, besides that there are many additional problems; simply, Terra's type safety is inadequate to make computed gotos not be an enormous landmine. Maybe that isn't a big problem.

Some cursory research regarding computed gotos in LLVM suggests that LLVM's jump table implementation might already have the optimizations which by their absence would make directly using a computed goto useful. So, that should probably be tested before deciding on computed goto versus jump table. Personally, jump table is the nicer abstraction most of the time, as computed goto would require rolling my own jump tables with no speedup for most common usecases, and even if computed gotos are available, I'd much rather also have a jump table and switch statement available.

**Comment by OvermindDL1** [comment]
> Some cursory research regarding computed gotos in LLVM suggests that LLVM's jump table implementation might already have the optimizations which by their absence would make directly using a computed goto useful.

So it can magically somehow perform a precaching step to, say, iterate over opcode and convert them directly to their jump addresses then a goto can jump directly to it?  It's that step of figuring out 'which' address to use that can have quite a sizeable overhead, especially if it involves many function calls (such as with the python interpreter).  Somehow I doubt that?

**Comment by elliottslaughter** [comment]
I agree that we should actually test this, if we believe it could be an issue, rather than debating it in the absence of evidence.

**Comment by ErikMcClure** [comment]
Unless someone is willing to provide an alternative implementation of jump tables, this language feature will be killed by bikeshedding.

At this point, I really don't care if some people think a switch statement is not the best possible solution for Terra unless they are willing to **actually implement an alternative**.

Terra needs some way of generating jump tables. This pull request provides a *fully functioning simplistic switch statement*, ready to be merged into master. I don't care how many people stand around saying "Well theoretically, it'd be better if we do X" if nobody's going to actually do anything. It's been 22 days now - you can "theoretically" until the heat death of the universe, but it won't mean anything if nobody **actually writes any code**.

If people want jump tables to potentially be a future language construct, we should remove the `do` part of the switch statement, as suggested by @elliottslaughter and simply not worry about duff's device. This allows people to do what they actually want, which is manually construct a switch statement on a bunch of constant integers. This covers 95% of actual use-cases and should be included in the language _regardless of any future features_ for the sake of usability.

The syntax modification would be a trivial change and the rest of the pull request could simply be merged as-is.

**Comment by aiverson** [comment]
We should probably include some variant of the macro we made to test the feature to provide the metaprogramming in terralib.

**Comment by elliottslaughter** [comment]
I think a switch statement is still valuable for ease of use and for familiarity coming from other languages---even if we also intend to add jump tables or computed goto in some other form.

Would it address the metaprogramming concerns if we allowed `case` statements to be metaprogrammed? I.e. allow something like:

```
local first = quote
  case 1 do
    ...
  end
end
local second = quote
  case [compute_something(...)] do
    ...
  end
end
local result = quote
  switch [some_expression(...)]
    [first]
    [second]
    else
      ...
  end
```

There would presumably be a restriction that `case` blocks must occur only at the top level of a switch block. The main issue in this syntax is that any escaped `case`s must appear first (i.e. they'd get sucked into another `case` if you put them behind one of the others).

If this is acceptable, I think we could take the PR in close to its current form, since this is a forward-compatible extension.

I agree that preserving the ability to generate optimized code is important, but it seems to me that the compiler has more than enough information to do any optimizations we may care about. Specifically the compiler has the values of all case expressions available at compile time, and if it were important, it could probably also get the addresses of the jumps. So I don't think we'll be limiting any future optimizations by taking the feature in its current form, and jump tables or computed goto are mostly a matter of what is convenience, not expressivity or performance.

**Comment by ErikMcClure** [comment]
This form of metaprogramming is already possible in the proposed pull request, since case statements are treated as a quoteable block just like everything else. This should include the ability to conditionally metaprogram the `else` block as well, and I have tested computing a constant integer in a case statement. So long as the constant calculation folds into a single integer constant, LLVM will accept it. If it doesn't, a compiler error is emitted explaining what went wrong.

We can augment the switch statement tests to include metaprogramming so we know just how much of the switch statement is metaprogrammable - but unless I am mistaken, every single part of it can be manually assembled by returning quotes.

**Comment by elliottslaughter** [comment]
I think what I'm suggesting is a restriction relative to the current PR, since that version was intended for eventually being extended to handle Duff's Device, which would require nesting `case` blocks inside other blocks. For the initial support I think it's best to require `case` to be nested directly under the `switch` block.

**Comment by aiverson** [comment]
The current PR does that already.

**Comment by aiverson** [comment]
Also, case statements aren't a quotable block in the current code, I believe. They can only occur syntactically immediately nested inside the switch statement. Allowing cases to be quoted would work, but it creates a big parsing problem to attempt to allow both quoted cases and cases that create scopes.

**Comment by elliottslaughter** [comment]
I've updated the RFC text above to explain how metaprogramming would work. I think this is sufficient to enable the basic use case, which again I think is valuable even if we subsequently add a more powerful construct like jump tables or computed goto.

@zdevito Does this version satisfy your concerns?

**Comment by elliottslaughter** [comment]
Updated the RFC to add `do` after `switch` back, both to make it more consistent with other control flow constructs (now that `case` is more standalone), and also to avoid hazards with `[]`. Thanks @aiverson for pointing this out.

**Comment by elliottslaughter** [comment]
Merged via #416. I'm going to consider this feature experimental while we see how it feels, but it available in master.
